lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,642 @@
+"""Agent policy for LP trading that also arbitrage on the fixed rate."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from fixedpointmath import FixedPoint
+
+from agent0.hyperdrive.state import HyperdriveActionType, HyperdriveMarketAction
+from elfpy.types import MarketType, Trade
+
+from .hyperdrive_policy import HyperdrivePolicy
+
+if TYPE_CHECKING:
+    from ethpy.hyperdrive import HyperdriveInterface
+    from numpy.random._generator import Generator as NumpyGenerator
+
+    from agent0.hyperdrive.state import HyperdriveWallet
+
+import time
+
+# pylint: disable=too-many-parameters, too-many-local-variables
+
+# constants
+fp0 = FixedPoint(0)
+fp1 = FixedPoint(1)
+fp2 = FixedPoint(2)
+fp12 = FixedPoint(12)
+fp_seconds_in_year = FixedPoint(365 * 24 * 60 * 60)
+tolerance = 1e-18
+MAX_ITER = 10
+MIN_TRADE_AMOUNT = FixedPoint(scaled_value=20)
+
+
+# functions
+def calc_bond_reserves(
+    share_reserves: FixedPoint,
+    share_price: FixedPoint,
+    target_rate: FixedPoint,
+    position_duration: FixedPoint,
+    time_stretch: FixedPoint,
+):
+    """Calculate the amount of bonds that hit the target target rate for the given shares.
+
+    Parameters
+    ----------
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_price : FixedPoint
+        The price of the share.
+    target_rate : FixedPoint
+        The target rate.
+    position_duration : FixedPoint
+        The duration of the position.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The amount of bonds that hit the target rate.
+    """
+    return share_price * share_reserves * ((fp1 + target_rate * position_duration / fp_seconds_in_year) ** time_stretch)
+
+
+def calc_spot_price_local(
+    initial_share_price: FixedPoint,
+    share_reserves: FixedPoint,
+    share_adjustment: FixedPoint,
+    bond_reserves: FixedPoint,
+    time_stretch: FixedPoint,
+) -> FixedPoint:
+    """Calculate spot price.
+
+    Parameters
+    ----------
+    initial_share_price : FixedPoint
+        The initial price of the share.
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_adjustment : FixedPoint
+        The amount of share adjustment.
+    bond_reserves : FixedPoint
+        The amount of bond reserves.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The spot price.
+    """
+    effective_share_reserves = share_reserves - share_adjustment
+    return (initial_share_price * effective_share_reserves / bond_reserves) ** time_stretch
+
+
+def calc_apr(
```

Comment:
Similar comment to above. This should be `calc_apr_local`, which I assume does not want to use the current pool state. we can make this a TODO here, but we should have an API function `calc_fixed_rate(spot_price)` that returns `(FixedPoint(1) - spot_price) / (spot_price * self.position_duration_in_years)` and then the property would be:
```python
@property
def fixed_rate(self) -> FixedPoint:
    """Returns the market fixed rate.

    Follows the formula:

    .. math::
        r = ((1/p)-1)/t = (1-p)/(pt)
    """
    return self.calc_fixed_rate(self.spot_price)
```